### ISSUE TITLE
Explore custom DB for ABCI

### ISSUE THREAD

**Issue by gchaincl** [ParadigmFoundation/ParadigmCore#60]
Tendermint uses `leveldb` as the main database backend
```toml
# Database backend: leveldb | memdb | cleveldb
db_backend = "leveldb"
```
We want to explore how feasible would it be to use a custom database engine.

**Comment by gchaincl** [comment]
Tendermint's code is designed to support custom _Backend Database_ injection. The implementation needs to satisfy `types.DB`[[1]](https://github.com/tendermint/tendermint/blob/e0f8936455029a40287a69d5b0e7baa4d5864da1/libs/db/types.go) interface. Doing so is straight forward, as we can see in the [memory implementation](https://github.com/tendermint/tendermint/blob/e0f8936455029a40287a69d5b0e7baa4d5864da1/libs/db/mem_db.go).

It's important to consider that running `tendermint node` will only support `leveldb | memdb | cleveldb`, if we want to customize our DB, we will need to use the in-process ABCI model.

The _Backend Database_ is responsible for storing the blockchain and its state data [[2]](https://tendermint.com/docs/tendermint-core/running-in-production.html#database). Note that this has nothing to do with the App State.
To store the AppState we can use whatever it fit out needs, in this [example](https://gist.github.com/gchaincl/d051aa8fade4bb433782d43fc5282a05#file-main-go-L32) we are using an IAVL tree with an in-memory database. IAVL supports injecting any database as long as it satisfies  [[1]](https://github.com/tendermint/tendermint/blob/e0f8936455029a40287a69d5b0e7baa4d5864da1/libs/db/types.go)

In summary:
The _Backend Database_ stores the blockchain data, it uses `leveldb` we can inject a custom one as long as we follow the in-process model.

The AppState uses a different database, and it's likely that we'll use a Merkle tree if we use IAVL it supports custom DB injection. There's no need to do in-process ABCI for this.

1: https://github.com/tendermint/tendermint/blob/e0f8936455029a40287a69d5b0e7baa4d5864da1/libs/db/types.go
2: https://tendermint.com/docs/tendermint-core/running-in-production.html#database

**Comment by gchaincl** [comment]
A SQLite3 implementation has been implemented for reference:
https://gist.github.com/gchaincl/072c898d08f130133369b15997aca348

**Comment by hrharder** [comment]
Great, so given all of this, what do you think is the sensible route for storing the IAVL tree in a DB? Given that we want to support [crash recovery of the most recently committed state](https://tendermint.com/docs/spec/abci/apps.html#crash-recovery) some persistent DB over a memory DB seems like a requirement. Do you see advantages to using levelDB as tendermint does?

Very cool SQL impl., is there a reason you chose to demo SQL? My knowledge of DB tradeoffs is limited, so I'm curious to hear your thoughts. At a bare minimum, we want the Golang ABCI application to support crash recovery from a state committed to disk. (note that in ParadigmCore as of v0.8, the state is kept in-memory so if a node dies/crashes/restarts in a live network, it has to re-sync the entire chain from genesis by replaying all transactions, rather than only the ones that occurred since the crash.)

**Comment by gchaincl** [comment]
I wouldn't replace the default leveldb without understanding 1) the volume of data that we will handle and 2) how we perform under some stress test / benchmarking (considering the volume). Documentation also suggests using the C-implementation of leveldb when running in production: https://tendermint.com/docs/tendermint-core/running-in-production.html#database, which I'd say should be trivial to do with the knowledge we've built.
For the App State, I'd apply the same reasoning (1 and 2) before going with something else from leveldb.

Good question, I chose SQLite to build a reference implementation (the minimum to make Tendermint's test pass), SQLite has no dependencies so it's easy to set up.
One of the important takeaways from this experiment is that we need to implement an Iterator; it returns an ordered subset of the keys. Implementing the Iterator thus will require some sort of indexing on the DB engine.

**Comment by hrharder** [comment]
Great points. Looking forward to discussing more.
